File: xcomponentlib/src/main/java/com/zhxh/xcomponentlib/arcibrary/CArcProgress.java
package com.zhxh.xcomponentlib.arcibrary;

import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.Rect;
import android.graphics.RectF;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.widget.ProgressBar;

import com.zhxh.xcomponentlib.R;

/**
 * Created by zhxh on 2020/4/11
 */

public class CArcProgress extends ProgressBar {
    public static final int STYLE_TICK = 1;
    public static final int STYLE_ARC = 0;
    private final int DEFAULT_LINEHEIGHT = dp2px(15);
    private final int DEFAULT_mTickWidth = dp2px(2);
    private final int DEFAULT_mRadius = dp2px(72);
    private final int DEFAULT_mUnmProgressColor = 0xffeaeaea;
    private final int DEFAULT_mProgressColor = Color.YELLOW;
    private final float DEFAULT_OFFSETDEGREE = 60;
    private final int DEFAULT_DENSITY = 4;
    private final int MIN_DENSITY = 2;
    private final int MAX_DENSITY = 8;
    private int mStyleProgress = STYLE_ARC;
    private boolean mBgShow;
    private float mRadius;
    private int mArcbgColor;
    private int mBoardWidth;
    private float mDegree = DEFAULT_OFFSETDEGREE;
    private RectF mArcRectF;
    private Paint mLinePaint;
    private Paint mArcPaint;
    private Paint mTextPaint;

    private int mUnmProgressColor;
    private int mProgressColor;
    private int mTickWidth;
    private float mTickDensity;
    private Bitmap mCenterBitmap;
    private Canvas mCenterCanvas;
    private OnCenterDraw mOnCenter;

    public CArcProgress(Context context) {
        this(context, null);
    }

    public CArcProgress(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CArcProgress(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        final TypedArray attributes = getContext().obtainStyledAttributes(
                attrs, R.styleable.CArcProgress);
        mBoardWidth = attributes.getDimensionPixelOffset(R.styleable.CArcProgress_borderWidth, DEFAULT_LINEHEIGHT);
        mUnmProgressColor = attributes.getColor(R.styleable.CArcProgress_unprogresColor, DEFAULT_mUnmProgressColor);
        mProgressColor = attributes.getColor(R.styleable.CArcProgress_progressColor, DEFAULT_mProgressColor);
        mTickWidth = attributes.getDimensionPixelOffset(R.styleable.CArcProgress_tickWidth, DEFAULT_mTickWidth);
        mTickDensity = attributes.getFloat(R.styleable.CArcProgress_tickDensity, DEFAULT_DENSITY);
        mRadius = attributes.getDimensionPixelOffset(R.styleable.CArcProgress_radius, DEFAULT_mRadius);
        mArcbgColor = attributes.getColor(R.styleable.CArcProgress_arcbgColor, DEFAULT_mUnmProgressColor);
        mTickDensity = Math.max(Math.min(mTickDensity, MAX_DENSITY), MIN_DENSITY);
        mBgShow = attributes.getBoolean(R.styleable.CArcProgress_bgShow, false);
        mDegree = attributes.getFloat(R.styleable.CArcProgress_degree, DEFAULT_OFFSETDEGREE);
        mStyleProgress = attributes.getInt(R.styleable.CArcProgress_progressStyle, STYLE_ARC);
        boolean capRount = attributes.getBoolean(R.styleable.CArcProgress_arcCapRound, false);
        mArcPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mArcPaint.setColor(mArcbgColor);
        if (capRount)
            mArcPaint.setStrokeCap(Paint.Cap.ROUND);
        mArcPaint.setStrokeWidth(mBoardWidth);
        mArcPaint.setStyle(Paint.Style.STROKE);
        mLinePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mLinePaint.setStrokeWidth(mTickWidth);

        mTextPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mTextPaint.setTextSize(12);
        mTextPaint.setColor(Color.parseColor("#999999"));
    }

    public void setOnCenterDraw(OnCenterDraw mOnCenter) {
        this.mOnCenter = mOnCenter;
    }

    @Override
    protected synchronized void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int widthMode = MeasureSpec.getMode(widthMeasureSpec);
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);
        if (widthMode != MeasureSpec.EXACTLY) {
            int widthSize = (int) (mRadius * 2 + mBoardWidth * 2);
            widthMeasureSpec = MeasureSpec.makeMeasureSpec(widthSize, MeasureSpec.EXACTLY);
        }
        if (heightMode != MeasureSpec.EXACTLY) {
            int heightSize = (int) (mRadius * 2 + mBoardWidth * 2);
            heightMeasureSpec = MeasureSpec.makeMeasureSpec(heightSize, MeasureSpec.EXACTLY);
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    protected synchronized void onDraw(Canvas canvas) {
        canvas.save();
        float rotate = getProgress() * 1.0f / getMax();
        float x = mArcRectF.right / 2 + mBoardWidth / 2;
        float y = mArcRectF.right / 2 + mBoardWidth / 2;
        if (mOnCenter != null) {
            if (mCenterCanvas == null) {
                mCenterBitmap = Bitmap.createBitmap((int) mRadius * 2, (int) mRadius * 2, Bitmap.Config.ARGB_8888);
                mCenterCanvas = new Canvas(mCenterBitmap);
            }
            mCenterCanvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR);
            mOnCenter.draw(mCenterCanvas, mArcRectF, x, y, mBoardWidth, getProgress());
            canvas.drawBitmap(mCenterBitmap, 0, 0, null);
        }
        float angle = mDegree / 2;
        int count = (int) ((360 - mDegree) / mTickDensity);
        int target = (int) (rotate * count);
        if (mStyleProgress == STYLE_ARC) {
            float targetDegree = (360 - mDegree) * rotate;
            //绘制完成部分
            mArcPaint.setColor(mProgressColor);

            canvas.drawArc(mArcRectF, 90 + angle, targetDegree, false, mArcPaint);
            //绘制未完成部分
            mArcPaint.setColor(mUnmProgressColor);
            canvas.drawArc(mArcRectF, 90 + angle + targetDegree, 360 - mDegree - targetDegree, false, mArcPaint);

            //绘制文字
            canvas.rotate(180 + angle, x, y);
            for (int i = 0; i <= count + 1; i = i + 10) {
                String text = String.valueOf(i);
                Rect textBound = new Rect();
                mTextPaint.getTextBounds(text, 0, text.length(), textBound);
                canvas.drawText(text, x, mBoardWidth + mBoardWidth / 2 + textBound.width(), mTextPaint);
                canvas.rotate(mTickDensity * 10, x, y);
            }
        } else {//钟表模式
            if (mBgShow)
                canvas.drawArc(mArcRectF, 90 + angle, 360 - mDegree, false, mArcPaint);
            canvas.rotate(180 + angle, x, y);
            for (int i = 0; i < count; i++) {
                if (i < target) {
                    mLinePaint.setColor(mProgressColor);
                } else {
                    mLinePaint.setColor(mUnmProgressColor);
                }
                canvas.drawLine(x, mBoardWidth + mBoardWidth / 2, x, mBoardWidth - mBoardWidth / 2, mLinePaint);
                canvas.rotate(mTickDensity, x, y);
            }
        }
        canvas.restore();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        mArcRectF = new RectF(mBoardWidth,
                mBoardWidth,
                mRadius * 2 - mBoardWidth,
                mRadius * 2 - mBoardWidth);
        Log.e("DEMO", "right == " + mArcRectF.right + "   mRadius == " + mRadius * 2);
    }

    /**
     * dp 2 px
     *
     * @param dpVal
     */
    protected int dp2px(int dpVal) {
        return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP,
                dpVal, getResources().getDisplayMetrics());
    }

    public interface OnCenterDraw {
        /**
         * @param canvas
         * @param rectF       圆弧的Rect
         * @param x           圆弧的中心x
         * @param y           圆弧的中心y
         * @param strokeWidth 圆弧的边框宽度
         * @param progress    当前进度
         */
        public void draw(Canvas canvas, RectF rectF, float x, float y, float strokeWidth, int progress);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if (mCenterBitmap != null) {
            mCenterBitmap.recycle();
            mCenterBitmap = null;
        }
    }

    public void runProgress(int progress, long itemDuration) {
        ValueAnimator va = ValueAnimator.ofInt(0, progress);
        va.setDuration(itemDuration * progress);
        va.addUpdateListener(valueAnimator -> {
            Integer p = (Integer) valueAnimator.getAnimatedValue();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                CArcProgress.this.setProgress(p, true);
            } else {
                CArcProgress.this.setProgress(p);
            }
        });
        va.start();
    }
}

File: xcomponentlib/src/main/java/com/zhxh/xcomponentlib/CArcProgress.kt
package com.zhxh.xcomponentlib

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.*
import android.os.Build
import android.util.AttributeSet
import android.util.TypedValue
import android.widget.ProgressBar

/**
 * Created by zhxh on 2020/4/14
 */
class CArcProgress @JvmOverloads constructor(context: Context?, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : ProgressBar(context, attrs, defStyleAttr) {
    private val DEFAULT_LINEHEIGHT = dp2px(12)
    private val DEFAULT_mTickWidth = dp2px(2)
    private val DEFAULT_mRadius = dp2px(72)
    private val DEFAULT_mUnmProgressColor = -0x151516
    private val DEFAULT_mProgressColor = Color.YELLOW
    private val DEFAULT_OFFSETDEGREE = 60f
    private val DEFAULT_DENSITY = 4
    private val MIN_DENSITY = 2
    private val MAX_DENSITY = 8
    private var mStyleProgress = STYLE_ARC
    private val mBgShow: Boolean
    private val mRadius: Float
    private val mArcbgColor: Int
    private val mBoardWidth: Int
    private var mDegree = DEFAULT_OFFSETDEGREE
    private var mArcRectF: RectF? = null
    private val mLinePaint: Paint
    private val mArcPaint: Paint
    private val mTextPaint: Paint
    private val mUnmProgressColor: Int
    private val mProgressColor: Int
    private val mTickWidth: Int
    private var mTickDensity: Float
    private var mCenterBitmap: Bitmap? = null
    private var mCenterCanvas: Canvas? = null
    private var mOnCenter: OnCenterDraw? = null

    private var animTim = 0L
    fun setOnCenterDraw(mOnCenter: OnCenterDraw?) {
        this.mOnCenter = mOnCenter
    }

    /**
     * progress 表示进度 0-100
     * plus增加了多少 例如 +3
     * rate 表示分数 例如 85分
     * defeat 例如 65.5%
     * itemDuration 速度 默认25ms
     */
    private fun runProgress(progress: Int, plus: String?, rate: String?, defeat: String, itemDuration: Long) {
        initDataDraw(plus, rate, defeat)
        startRun(progress, itemDuration)
    }

    /**
     * progress 表示进度 0-100
     * plus增加了多少 例如 +3
     * rate 表示分数 例如 85分
     * defeat 例如 65.5%
     */
    fun runProgress(progress: Int, plus: String?, rate: String?, defeat: String) {
        runProgress(progress, plus, rate, defeat, 25)
    }

    /**
     * 改函数适用当没有plus情况
     */
    fun runProgress(progress: Int, rate: String?, defeat: String) {
        runProgress(progress, "", rate, defeat)
    }

    private fun initDataDraw(plus: String?, rate: String?, defeat: String) {
        setOnCenterDraw(object : OnCenterDraw {
            override fun draw(canvas: Canvas?, rectF: RectF?, x: Float, y: Float, strokeWidth: Float, progress: Int) {
                if (canvas == null) {
                    return
                }
                if (rectF == null) {
                    return
                }

                run {
                    val textPaint = Paint(Paint.ANTI_ALIAS_FLAG)
                    if (!plus.isNullOrEmpty()) {
                        textPaint.color = Color.parseColor("#999999")
                        textPaint.textSize = dp2px(12).toFloat()
                        canvas.drawText(plus, x - textPaint.measureText(plus) / 2, rectF.top + dp2px(50), textPaint)
                    }
                }

                val cancelRun = Runnable {
                    val textPaint = Paint(Paint.ANTI_ALIAS_FLAG)
                    textPaint.textSize = dp2px(12).toFloat()

                    val paint = Paint()
                    paint.color = Color.WHITE
                    canvas.drawRect(x - textPaint.measureText(plus) / 2, rectF.top + dp2px(50), x + textPaint.measureText(plus) / 2, rectF.top + dp2px(35), paint)
                }

                handler.postDelayed(cancelRun, animTim + 250)

                val textPaint = Paint(Paint.ANTI_ALIAS_FLAG)
                textPaint.flags = Paint.FAKE_BOLD_TEXT_FLAG
                textPaint.color = Color.parseColor("#333333")
                textPaint.textSize = dp2px(28).toFloat()
                val progressStr = "${progress}分"
                canvas.drawText(progressStr, x - textPaint.measureText(progressStr) / 2, rectF.top + dp2px(80), textPaint)

                if (!rate.isNullOrEmpty()) {
                    textPaint.flags = Paint.ANTI_ALIAS_FLAG
                    textPaint.color = Color.parseColor("#333333")
                    textPaint.textSize = dp2px(18).toFloat()
                    canvas.drawText(rate, x - textPaint.measureText(rate) / 2, rectF.top + dp2px(120), textPaint)
                }

                textPaint.flags = Paint.ANTI_ALIAS_FLAG
                textPaint.color = Color.parseColor("#999999")
                textPaint.textSize = dp2px(15).toFloat()

                val start = "- 打败了"
                val end = "的在投用户 -"
                val percentStr = "$start$defeat$end"

                val start1X = x - textPaint.measureText(percentStr) / 2
                val start2X = start1X + textPaint.measureText(start)
                val start3X = start2X + textPaint.measureText(defeat)

                canvas.drawText(start, start1X, rectF.bottom - 10, textPaint)
                canvas.drawText(end, start3X, rectF.bottom - 10, textPaint)
                textPaint.color = Color.parseColor("#FF403B")
                canvas.drawText(defeat, start2X, rectF.bottom - 10, textPaint)
            }
        })
    }

    @Synchronized
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var widthMeasureSpec = widthMeasureSpec
        var heightMeasureSpec = heightMeasureSpec
        val widthMode = MeasureSpec.getMode(widthMeasureSpec)
        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        if (widthMode != MeasureSpec.EXACTLY) {
            val widthSize = (mRadius * 2 + mBoardWidth * 2).toInt()
            widthMeasureSpec = MeasureSpec.makeMeasureSpec(widthSize, MeasureSpec.EXACTLY)
        }
        if (heightMode != MeasureSpec.EXACTLY) {
            val heightSize = (mRadius * 2 + mBoardWidth * 2).toInt()
            heightMeasureSpec = MeasureSpec.makeMeasureSpec(heightSize, MeasureSpec.EXACTLY)
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    @Synchronized
    override fun onDraw(canvas: Canvas) {
        canvas.save()
        val rotate = progress * 1.0f / max
        val x = mArcRectF!!.right / 2 + mBoardWidth / 2
        val y = mArcRectF!!.right / 2 + mBoardWidth / 2
        if (mOnCenter != null) {
            if (mCenterCanvas == null) {
                mCenterBitmap = Bitmap.createBitmap(mRadius.toInt() * 2, mRadius.toInt() * 2, Bitmap.Config.ARGB_8888)
                mCenterCanvas = Canvas(mCenterBitmap)
            }
            mCenterCanvas!!.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
            mOnCenter!!.draw(mCenterCanvas, mArcRectF, x, y, mBoardWidth.toFloat(), progress)
            canvas.drawBitmap(mCenterBitmap, 0f, 0f, null)
        }
        val angle = mDegree / 2
        val count = ((360 - mDegree) / mTickDensity).toInt()
        val target = (rotate * count).toInt()
        if (mStyleProgress == STYLE_ARC) {
            val targetDegree = (360 - mDegree) * rotate
            //绘制未完成部分
            mArcPaint.color = mUnmProgressColor
            canvas.drawArc(mArcRectF, 90 + angle + targetDegree, 360 - mDegree - targetDegree, false, mArcPaint)
            //绘制完成部分
            mArcPaint.color = mProgressColor
            canvas.drawArc(mArcRectF, 90 + angle, targetDegree, false, mArcPaint)
            //绘制文字
            canvas.rotate(180 + angle, x, y)
            var i = 0
            while (i <= count + 1) {
                val text = i.toString()
                val textBound = Rect()
                mTextPaint.getTextBounds(text, 0, text.length, textBound)
                canvas.drawText(text, x, mBoardWidth + mBoardWidth / 2 + textBound.height() * 1.5f, mTextPaint)
                canvas.rotate(mTickDensity * 10, x, y)
                i += 10
            }
            //绘制自定义线帽 while多旋转了20度
            val pointPaint = Paint(Paint.ANTI_ALIAS_FLAG)
            pointPaint.strokeWidth = mBoardWidth.toFloat()
            pointPaint.color = Color.WHITE
            canvas.rotate(180 + mTickDensity * target - angle - 20, x, y)
            canvas.drawCircle(x, (mBoardWidth).toFloat(), mBoardWidth.toFloat() / 3, pointPaint)

        } else { //钟表模式
            if (mBgShow) canvas.drawArc(mArcRectF, 90 + angle, 360 - mDegree, false, mArcPaint)
            canvas.rotate(180 + angle, x, y)
            for (i in 0 until count) {
                if (i < target) {
                    mLinePaint.color = mProgressColor
                } else {
                    mLinePaint.color = mUnmProgressColor
                }
                canvas.drawLine(x, mBoardWidth + mBoardWidth / 2.toFloat(), x, mBoardWidth - mBoardWidth / 2.toFloat(), mLinePaint)
                canvas.rotate(mTickDensity, x, y)
            }
        }
        canvas.restore()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        mArcRectF = RectF(mBoardWidth.toFloat(),
                mBoardWidth.toFloat(),
                mRadius * 2 - mBoardWidth,
                mRadius * 2 - mBoardWidth)
    }

    /**
     * dp 2 px
     *
     * @param dpVal
     */
    private fun dp2px(dpVal: Int): Int {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP,
                dpVal.toFloat(), resources.displayMetrics).toInt()
    }

    interface OnCenterDraw {
        /**
         * @param canvas
         * @param rectF       圆弧的Rect
         * @param x           圆弧的中心x
         * @param y           圆弧的中心y
         * @param strokeWidth 圆弧的边框宽度
         * @param progress    当前进度
         */
        fun draw(canvas: Canvas?, rectF: RectF?, x: Float, y: Float, strokeWidth: Float, progress: Int)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        if (mCenterBitmap != null) {
            mCenterBitmap!!.recycle()
            mCenterBitmap = null
        }
    }

    fun startRun(progress: Int, itemDuration: Long) {
        val va = ValueAnimator.ofInt(0, progress)
        animTim = itemDuration * progress
        va.duration = animTim
        va.addUpdateListener { valueAnimator: ValueAnimator ->
            val p = valueAnimator.animatedValue as Int
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                this@CArcProgress.setProgress(p, true)
            } else {
                this.progress = p
            }
        }
        va.start()
    }

    companion object {
        const val STYLE_TICK = 1
        const val STYLE_ARC = 0
    }

    init {
        val attributes = getContext().obtainStyledAttributes(
                attrs, R.styleable.CArcProgress)
        mBoardWidth = attributes.getDimensionPixelOffset(R.styleable.CArcProgress_borderWidth, DEFAULT_LINEHEIGHT)
        mUnmProgressColor = attributes.getColor(R.styleable.CArcProgress_unprogresColor, DEFAULT_mUnmProgressColor)
        mProgressColor = attributes.getColor(R.styleable.CArcProgress_progressColor, DEFAULT_mProgressColor)
        mTickWidth = attributes.getDimensionPixelOffset(R.styleable.CArcProgress_tickWidth, DEFAULT_mTickWidth)
        mTickDensity = attributes.getFloat(R.styleable.CArcProgress_tickDensity, DEFAULT_DENSITY.toFloat())
        mRadius = attributes.getDimensionPixelOffset(R.styleable.CArcProgress_radius, DEFAULT_mRadius).toFloat()
        mArcbgColor = attributes.getColor(R.styleable.CArcProgress_arcbgColor, DEFAULT_mUnmProgressColor)
        mTickDensity = Math.max(Math.min(mTickDensity, MAX_DENSITY.toFloat()), MIN_DENSITY.toFloat())
        mBgShow = attributes.getBoolean(R.styleable.CArcProgress_bgShow, false)
        mDegree = attributes.getFloat(R.styleable.CArcProgress_degree, DEFAULT_OFFSETDEGREE)
        mStyleProgress = attributes.getInt(R.styleable.CArcProgress_progressStyle, STYLE_ARC)
        val capRound = attributes.getBoolean(R.styleable.CArcProgress_arcCapRound, false)
        mArcPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mArcPaint.color = mArcbgColor
        if (capRound) mArcPaint.strokeCap = Paint.Cap.ROUND
        mArcPaint.strokeWidth = mBoardWidth.toFloat()
        mArcPaint.style = Paint.Style.STROKE
        mLinePaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mLinePaint.strokeWidth = mTickWidth.toFloat()
        mTextPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mTextPaint.textSize = dp2px(8).toFloat()
        mTextPaint.color = Color.parseColor("#999999")
    }
}